### ISSUE TITLE
Reclaim keypress events from vispy, and clarify what actually handles keypresses.

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#3087]
## 🧰 Task

In looking through some of the keypress handling, I think we would benefit from reclaiming the handling of keypress events from the vispy canvas, and generally clarifying _what_ handles keypress events.

Currently, most keypress events that happen will either be directly handled by the vispy canvas (if the canvas has focus), or will be handled by QtViewer.  This makes sense because Qt passed Key events [to the widget with keyboard input focus when keys are pressed or released](https://doc.qt.io/qt-5/qkeyevent.html#details).  However, we seem to be routing _all_ of our keypress events through vispy's canvas, regardless of whether it has keyboard focus or not and regardless of whether the keypress is acting on the canvas.  Perhaps this was because historically it was the first widget that people wanted to connect key press events to?  The relevant code starts here:

https://github.com/napari/napari/blob/258f32d9c960e054bd3e58f5797a63b7b8874a01/napari/_qt/qt_viewer.py#L886-L895

...where the QtViewer essentially passes on all keypresses, and defers to a private method on the vispy canvas's (private) backend.  (That `_keyEvent` method, btw, calls [this function in vispy](https://github.com/vispy/vispy/blob/main/vispy/app/backends/_qt.py#L562-L572), converting keys into vispy's internal representation).

We then, however, _re_-claim control of canvas keypress events by connecting the canvas events back to the QtViewer's `_key_map_handler` object here:

https://github.com/napari/napari/blob/258f32d9c960e054bd3e58f5797a63b7b8874a01/napari/_qt/qt_viewer.py#L278-L279

> *side-note*: the only reason that works for `key_press` is because it uses vispy's magic function name inspection to connect `on_<function_name>` to the appropriate event name... which we sometimes use and sometimes don't. I think we should generally avoid it.

So a keypress on the viewer results in:

1. `QtViewer.keyPressEvent` > 
2. `vispy.app.backends._qt.QtBaseCanvasBackend._keyEvent` > 
3. `vispy.scene.canvas.SceneCanvas.events.key_press` > 
4. `QtViewer._key_map_handler.on_key_press` > 
5. `QtViewer._key_map_handler.press_key` 

... (and `press_key` actually executes the callback).  I think we should remove this indirection, and make `QtViewer.keyPressEvent` itself dispatch to some action manager, removing the vispy intermediary, and preferably also removing the `KeymapHandler`... which ought to be the ActionManager at this point

**Comment by sofroniewn** [comment]
>  I think we would benefit from reclaiming the handling of keypress events from the vispy canvas, and generally clarifying what handles keypress events.

Agreed! I'd be supportive of this. Could be something good for @ppwadhwa to take on

**Comment by tlambert03** [comment]
I'm starting to work on this... just adding some notes here on how this currently works for personal future reference:

### current keypress handling

#### if a key is pressed When qt-viewer is focused...

1. `QtViewer.keyPressEvent` retrieves the OS key press event
2. that calls `vispy.app.backends._qt.CanvasBackendDesktop._keyEvent` (`QtViewer.canvas._backend`)
    - this performs conversion of native Qt QKeyEvent to vispy key/modifiers
3. that emits `QtViewer.canvas.events.key_press(native=..., key=..., modifiers=...)`  (event emission)
4. that triggers: `KeymapHandler.on_key_press`
    - (connected in `QtViewer._create_canvas` with `self.canvas.events.key_press.connect(self._key_map_handler.on_key_press)`)
    - another conversion happens, from vispy key/modifiers to pure string form with `components_to_key_combo`
5. that calls `KeymapHandler.press_key(combo)` where `combo` is a pure string
    - `combo` is again processed with `normalize_key_combo(combo)`
    - that then re-parses the combo via regex with `key, modifiers = parse_key_combo(combo)`
    - parsed key & modifiers are then converted back again to a string with `key_combo = components_to_key_combo(key, modifiers)`
    - that `key_combo` is then looked up in a ChainMap of `KeymapProvider` (this map is rebuilt from the current `KeymapHandler.keymap_providers` on every key press)
6. if a callback is found for the specific key_combo string, it is then executed.

#### if a key is pressed when `QtViewer.canvas` itself has focus...

then the receiver in step 1 above is replaced with `vispy.app.backends._qt.CanvasBackendDesktop` ... but then we go to step 2 and the rest is the same

#### if anything else has focus...

If it's the main window, nothing happens, keypresses aren't handled
if it's another widget... then depending on that widget's keypress handling, it may or may bubble up and be handled.


### proposed change

there's quite a lot of redirection and many conversion steps in there that could be removed.  I see two general options:

1. change the `QtViewer.keyPressEvent` to directly convert the QKeyEvent into an `app-model` keybinding, then trigger the corresponding action in the KeyBindingsRegistry
2. let Qt _itself_ just call the callback, by converting the prescribed key combo into a QAction/QShortcut that is registered with Qt.

the second option likely has the least amount of overhead... whereas the first might be a bit easier to reason about, particularly if the keybings are being dynamically modified by the user (though, thats also not incompatible with the second option... we'd just a "keybinding updated" event to update the QShortcuts/QActions)

**Comment by Czaki** [comment]
> change the `QtViewer.keyPressEvent` to directly convert the QKeyEvent into an `app-model` keybinding, then trigger the corresponding action in the KeyBindingsRegistry

You mean stop using vispy conversion with overwrite of `QtViewerDockWidget.keyPressEvent` to not accept event?

In your description you also commit case of Dock widget that calls `self._ref_qt_viewer().keyPressEvent(event)` to pass event to canvas.

**Comment by tlambert03** [comment]
> You mean stop using vispy conversion

yeah, no vispy in the loop at all.

> overwrite of QtViewerDockWidget.keyPressEvent to not accept event?

in what I've described so far, I haven't yet touched on dock widgets...  I'm just describing the current handling of keypress events on `QtViewer` or `QtViewer.canvas` ... will look at dock widgets later, but that's why I think we'd probably want to put our final keypress handler on QMainWindow (not on QtViewer)

**Comment by tlambert03** [comment]
and for reference... "reclaiming" the keypress events from vispy and simply redirecting them to QtViewer can look like this

```python
# napari/napari/_vispy/canvas.py
class VispyCanvas(SceneCanvas):
    ...

    # overriding keyPressEvent and keyReleaseEvent
    def create_native(self):
        super().create_native()
        self._backend.keyPressEvent = self._keyPressEvent
        self._backend.keyReleaseEvent = self._keyReleaseEvent

    def _keyPressEvent(self, event: QKeyEvent):
        print("keyPressEvent on canvas", event.key())
        # pass it through to qtviewer
        QWidget.keyPressEvent(self._backend, event)

    def _keyReleaseEvent(self, event: QKeyEvent):
        print("keyReleaseEvent on canvas", event.key())
        # pass it through to qtviewer
        QWidget.keyReleaseEvent(self._backend, event)
```

**Comment by Czaki** [comment]
You could use simple [`event.ignore()`](https://doc.qt.io/qt-6/qevent.html#ignore) and Qt will automatically call key handling in parent class. There is no need to call basic implementation from QWidget. 



> > overwrite of QtViewerDockWidget.keyPressEvent to not accept event?
> 
> in what I've described so far, I haven't yet touched on dock widgets

Wrong edition of my message. I was thinking about `QtBaseCanvasBackend` that you overwrite in your example if I correctly understand.

**Comment by tlambert03** [comment]
> You could use simple [event.ignore()](https://doc.qt.io/qt-6/qevent.html#ignore) and Qt will automatically call key handling in parent class. There is no need to call basic implementation from QWidget.

ok ... I did that because of this line in the [qt documentation](https://doc.qt.io/qt-6/qwidget.html#keyPressEvent):

> If you reimplement this handler, it is very important that you call the base class implementation if you do not act upon the key.

did I misunderstand it?

**Comment by Czaki** [comment]
> > If you reimplement this handler, it is very important that you call the base class implementation if you do not act upon the key.
> 
> did I misunderstand it?

Yes and no. Next paragraph is saying:

> The default implementation closes popup widgets if the user presses the key sequence for [QKeySequence::Cancel](https://doc.qt.io/qt-6/qkeysequence.html#StandardKey-enum) (typically the Escape key). Otherwise the event is ignored, so that the widget's parent can interpret it.

So call `.ignore()` do explicit what do default implementation in our case. 

But You do not call `keyEvent` of the base class but key event for QWidget. For the current code, it is exactly the same but it may change. This note is mainly for situations where you inherit from the class that already (or possibly could) handle keys for some events. 

There is a difference between: "Oh, I'm not interested in handling these keys, so call default behavior" and "I do not want to support this shortcut in this widget. Move it to parent object."

**Comment by Czaki** [comment]
To be more clear. Calling `event.ignore()` is acting upon the key. You handle the key and decide that its pressing should not trigger anything in this widget.

**Comment by tlambert03** [comment]
got it.  thanks!

**Comment by sofroniewn** [comment]
> yeah, no vispy in the loop at all.

Love this - only thing we need to then do is make sure all the camera functionality is implemented right - so zooming, panning, rotation etc. 

I think this will make some of the complex 3D interactivity stuff @brisvag @alisterburt @kevinyamauchi have been interested in much easier.

It will also mean that we don't have vispy objects flowing through our main APIs anymore, which I love as we can have greater consistency with the rest of napari and easier documentation
